### PR TITLE
Add balance function and tests

### PR DESCRIPTION
**Pull request by GoRunMaxim**

### CODE REVIEW

File: Balance_test.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		brackets   string
+		isBalanced bool
+	}{
+		{
+			brackets:   "{}()[]",
+			isBalanced: true,
+		},
+		{
+			brackets:   "{[(])({})}",
+			isBalanced: true,
```
Comment (papahmsolo) [inline Balance_test.go:16]: this test should fail

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Not only you should check the number of the various parentheses matches, but also check if the order is correct.

File: Balance.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// package balanced helps us to verifies strings by sequence of brackets
+package balanced
+
+// func Balance verifies if the given string is a balanced sequence of brackets
+func Balance(brackets string) bool {
+	var stack []byte
+	for _, v := range brackets {
+		if v == '{' || v == '(' || v == '[' {
+			stack = append(stack, byte(v))
```
Comment (papahmsolo) [inline Balance.go:9]: We won't need this conversion if we have a stack of `rune`

File: Balance.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+				}
+			case '[':
+				if v != ']' {
+					return false
+				}
+			case '(':
+				if v != ')' {
+					return false
+				}
+			}
```
Comment (papahmsolo) [inline Balance.go:29]: We can get rid of this long switch, try to store the matching brackets in a map.

File: Balance_test.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		brackets   string
+		isBalanced bool
+	}{
+		{
+			brackets:   "[]{}()[]",
+			isBalanced: true,
+		},
+		{
+			brackets:   "{[(])({})}",
+			isBalanced: false,
```
Comment (papahmsolo) [inline Balance_test.go:16]: Let's test more corner cases, like an empty string. double closing bracket in the end

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Nice job! You got the algorithm right!

File: Balance.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
+// package balanced helps us to verifies strings by sequence of brackets
+package balanced
+
+// func Balance verifies if the given string is a balanced sequence of brackets
+func Balance(brackets string) bool {
```
[the remark this answers is not in the document]
Comment (GoRunMaxim) [reply Balance.go:5]: correct, didn't set my ml settings

File: Balance.go (hunk 2 above)
Comment (GoRunMaxim) [reply Balance.go:9]: Fixed

File: Balance_test.go (hunk 4 above)
Comment (GoRunMaxim) [reply Balance_test.go:16]: Fixed

File: Balance.go (hunk 3 above)
Comment (GoRunMaxim) [reply Balance.go]: Fixed

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks great! 

Let's also make this work with expressions containing not only brackets, to validate things like this:
`((1 + 2) * 3) - 4)/5`

File: Balance.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+			if len(stack) == 0 {
+				return false
+			}
+			n := len(stack) - 1
+			last := map[rune]rune{
+				'{': '}',
+				'[': ']',
+				'(': ')',
+			}
+			i, ok := last[stack[n]]
```
[the remark this answers is not in the document]
Comment (GoRunMaxim) [reply Balance.go:20]: Nope. Deleted

File: Balance.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+		} else {
+			if len(stack) == 0 {
+				return false
+			}
+			n := len(stack) - 1
+			last := map[rune]rune{
+				'{': '}',
+				'[': ']',
+				'(': ')',
+			}
```
[the remark this answers is not in the document]
Comment (GoRunMaxim) [reply Balance.go:19]: ✅

File: Balance.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
+// Package balanced helps us to verifies strings by sequence of brackets
+package balanced
+
+var last = map[rune]rune{
+	'{': '}',
+	'[': ']',
+	'(': ')',
+}
```
Comment (papahmsolo) [inline Balance.go:8]: Isn't it a map of _matching_ brackets? I think the name is not self-descriptive here

File: Balance.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+// Balance func verifies if the given string is a balanced sequence of brackets
+func Balance(brackets string) bool {
+	var stack []rune
+	for _, v := range brackets {
+		if v == '{' || v == '(' || v == '[' {
+			stack = append(stack, v)
+		} else if v == '}' || v == ')' || v == ']' {
```
Comment (papahmsolo) [inline Balance.go:16]: If-else is fine, but `switch` is a bit cleaner

File: Balance.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+			if len(stack) == 0 {
+				return false
+			}
+			n := len(stack) - 1
+			i := last[stack[n]]
```
Comment (papahmsolo) [inline Balance.go:21]: `i` stands for index? Not sure about this name

File: Balance.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+			if i != v {
+				return false
+			}
```
Comment (papahmsolo) [inline Balance.go:24]: These `ifs` have the same return value, maybe we can combine them?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool, almost ready to move on

File: Balance.go (hunk 10 above)
Comment (GoRunMaxim) [reply Balance.go:21]: ✅

File: Balance.go (hunk 8 above)
Comment (GoRunMaxim) [reply Balance.go:8]: ✅

File: Balance.go (hunk 9 above)
Comment (papahmsolo) [reply Balance.go:16]: Try it out and we can compare then

File: Balance.go (hunk 11 above)
Comment (papahmsolo) [reply Balance.go]: Just combine two conditions into one if, simple as that

File: Balance.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+	for _, v := range brackets {
+		switch v {
+		case '{', '[', '(':
+			stack = append(stack, v)
+		case '}', ']', ')':
+			n := len(stack) - 1
+			if len(stack) == 0 || matchBrackets[stack[n]] != v {
+				return false
+			}
+			stack = stack[:n]
```
Comment (papahmsolo) [inline Balance.go:22]: I think the compiler can make that optimization with `n` for you. We write code for humans in the first place.
```			
if len(stack) == 0 || matchBrackets[stack[len(stack)-1]] != v {
    return false
}
stack = stack[:len(stack)-1]```

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Shouldn't file names be in lowercase?

File: Balance.go (hunk 12 above)
Comment (papahmsolo) [reply Balance.go:22]: This is optional, though

**Review by papahmsolo — APPROVED** [review]
